0017-ilp-errors/0017-ilp-errors.md

```diff
@@ -0,0 +1,73 @@
+| **S04** | **Insufficient Destination Amount** | The amount was insufficient (e.g. you tried to pay a $100 invoice with $10). |
+| **S05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. |
+| **S06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) |
+| **S07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint, e.g. a limit was exceeded. |
+
+**S08** - **S50** are reserved for future use.
+**S51** - **S99** are for application-defined errors.
+
+### T__ - Temporary Error
+
+Temporary errors indicate a failure on the part of the receiver or an intermediary system that is unexpected or likely to be resolved soon. Senders SHOULD retry the same payment again, possibly after a short delay.
+
+| Code | Name | Description |
+|---|---|---|
+| **T00** | **Internal Error** | Like HTTP 500. Something threw an exception, that's all we know. Try again later after we've had time to fix it. |
+| **T01** | **Ledger Unreachable** | The connector was not able to reach the next ledger. Try again later. |
+| **T02** | **Ledger Busy** | The ledger is rejecting requests due to overloading. Try again later. |
+| **T03** | **Connector Busy** | The connector is rejecting requests due to overloading. Try again later. |
+| **T04** | **Insufficient Liquidity** | The connector would like to fulfill your request, but it doesn't currently have enough money. Try again later. |
+
+**T05** - **T50** are reserved for future use.
+**T51** - **T99** are for application-defined errors.
+
+### R__ - Relative Error
+
+Relative errors indicate that the payment did not have enough of a margin in terms of money or time. However, it is impossible to tell whether the sender did not provide enough error margin or the path suddenly became too slow or illiquid. The sender MAY retry the payment with a larger safety margin.
+
+| Code | Name | Description
+|---|---|---|
+| **R01** | **Transfer Timed Out** | The transfer timed out, i.e. the next party in the chain did not respond. This could be because you set your timeout too low or because something look longer than it should. Try again with a higher expiry. |
+| **R02** | **Insufficient Source Amount** | Either you didn't send enough money or there wasn't enough liquidity. Try again with a higher sending amount. |
+| **R03** | **Insufficient Timeout** | The connector could not forward the payment, because the timeout was too low to subtract its safety margin. Try again with a higher expiry. |
+
+**R04** - **R50** are reserved for future use.
+**R51** - **R99** are for application-defined errors.
```

Comment:
Do we have a rule or registration process for keeping application-defined errors different from one another? With only 49 error codes to work with, I suspect it's very likely different applications will pick the same error codes to mean different things. With the way error codes are passed back through the chain, it seems like there's not much room for apps to namespace their own errors.

I don't think applications necessarily need to define their own top-level error codes if they have a structured field in the body where they can provide sub-codes, but maybe it would be helpful to have just one code set aside for application-defined errors.